api/client.go

```diff
@@ -1249,6 +1249,83 @@ START:
 	return result, nil
 }
 
+// requestWithContext avoids the use of the go-retryable library found in RawRequestWithContext and is
+// useful when making requests where the req or resp body is likely larger than available memory
+func (c *Client) requestWithContext(ctx context.Context, r *Request) (*Response, error) {
+	req, err := http.NewRequestWithContext(ctx, r.Method, r.URL.RequestURI(), r.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.URL.User = r.URL.User
+	req.URL.Scheme = r.URL.Scheme
+	req.URL.Host = r.URL.Host
+	req.Host = r.URL.Host
+
+	if r.Headers != nil {
+		for header, vals := range r.Headers {
+			for _, val := range vals {
+				req.Header.Add(header, val)
+			}
+		}
+	}
+
+	if len(r.ClientToken) != 0 {
+		req.Header.Set(consts.AuthHeaderName, r.ClientToken)
+	}
+
+	if len(r.WrapTTL) != 0 {
+		req.Header.Set("X-Vault-Wrap-TTL", r.WrapTTL)
+	}
+
+	if len(r.MFAHeaderVals) != 0 {
+		for _, mfaHeaderVal := range r.MFAHeaderVals {
+			req.Header.Add("X-Vault-MFA", mfaHeaderVal)
+		}
+	}
+
+	if r.PolicyOverride {
+		req.Header.Set("X-Vault-Policy-Override", "true")
+	}
+
+	var result *Response
+
+	resp, err := c.config.HttpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check for a redirect, only allowing for a single redirect
+	if resp.StatusCode == 301 || resp.StatusCode == 302 || resp.StatusCode == 307 {
+		// Parse the updated location
+		respLoc, err := resp.Location()
+		if err != nil {
+			return nil, err
+		}
+
+		// Ensure a protocol downgrade doesn't happen
+		if req.URL.Scheme == "https" && respLoc.Scheme != "https" {
+			return nil, fmt.Errorf("redirect would cause protocol downgrade")
+		}
+
+		// Update the request
+		req.URL = respLoc
+
+		// Retry the request
+		resp, err = c.config.HttpClient.Do(req)
```

Comment:
We don't need r.ResetJSONBody?

Reply:
The existing redirect logic in the snapshot writer omitted this, but It seems that we should!

Reply:
Does it?  I was basing this comment on comparing your code with RawRequestWithContext.

Reply:
I'm pretty sure that is what's going on [here](https://github.com/hashicorp/vault/blob/main/api/sys_raft.go#L189-L210).
I've refactored this so both the read and write snapshot functions will use this new client method and is a good opportunity to clean up some of this one-off code and get closer to the logic `RawRequestWithContext` has implemented.